src/main/java/graphql/execution/instrumentation/dataloader/DataLoaderDispatcherInstrumentationState.java

```diff
@@ -9,6 +12,14 @@
  */
 public class DataLoaderDispatcherInstrumentationState implements InstrumentationState {
 
+    @Internal
+    public static final DataLoaderRegistry EMPTY_DATALOADER_REGISTRY = new DataLoaderRegistry() {
+        @Override
+        public DataLoaderRegistry register(String key, DataLoader<?, ?> dataLoader) {
+            return Assert.assertShouldNeverHappen("You MUST set in your own DataLoaderRegistry to use data loader");
+        }
+    };
```

Comment:
Today a person could technically take the default DLR from the execution input and use to to register in dataloaders.

But I really want to retain the optimisation thatw e do no work when they have not set a DLR.

I figure we have 2 choices.  Use a null in the ExecutionInput and null check or use a default well know one and blow up.

I can see arguments for either way.  Thoughts?

Reply:
> But I really want to retain the optimisation thatw e do no work when they have not set a DLR.

So if I understand correctly, you are trying to optimize only for apps that do not use Dataloader **at all**. 
You are not trying to optimize for apps that have a mix of queries where some leverage data loader and others don't? (although I think it would be hard to know ahead of time which queries would trigger data loading vs not)

--
Also, the default seems reasonable to avoid NPEs downstream.

Reply:
We are trying to optimize for both ease of use AND performance

So We automatically add `DataLoaderDispatcherInstrumentation` to the mix if its not there

```
        boolean containsDLInstrumentation = instrumentationList.stream().anyMatch(instr -> instr instanceof DataLoaderDispatcherInstrumentation);
        if (!containsDLInstrumentation) {
            instrumentationList.add(new DataLoaderDispatcherInstrumentation());
        }
```

However if you dont use any DataLoaders then this can become a noop like thing so we cant be faster and no track all your fields and call dispatch on an empty registry.

if you look at `graphql.execution.instrumentation.dataloader.FieldLevelTrackingApproach` which is all the state we need to know what fields have completed and so on its non trivial

If your not using DataLoaders then you save memory and a little bit of CPU

Perhaps we should never have auto added `DataLoaderDispatcherInstrumentation` but we kept getting support request from the field saying data loaders dont work and we had to say "yeah you need to add a specvial `DataLoaderDispatcherInstrumentation` to make this work.

Reply:
> Perhaps we should never have auto added DataLoaderDispatcherInstrumentation but we kept getting support request from the field saying data loaders dont work and we had to say "yeah you need to add a specvial DataLoaderDispatcherInstrumentation to make this work.

Yea and in that case we could have validated at the time of execution of a request that a `DataLoaderDispatcherInstrumentation` must be present if an ExecutionInput specified a `DataLoaderRegistry`. That way people would get an error and just add the instrumentation right away instead of opening an issue. Ease of use (auto-configuration) vs performance is definitely a balance though.